pkg/ucp/api/v20231001preview/resourceprovider_conversion.go

```diff
@@ -0,0 +1,151 @@
+/*
+Copyright 2023 The Radius Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v20231001preview
+
+import (
+	v1 "github.com/radius-project/radius/pkg/armrpc/api/v1"
+	"github.com/radius-project/radius/pkg/to"
+	"github.com/radius-project/radius/pkg/ucp/datamodel"
+)
+
+const (
+	// ResourceProviderResourceType is the UCP fully-qualified type of the ResourceProviderResource resource.
+	ResourceProviderResourceType = "System.Resources/resourceProviders"
+)
+
+// ConvertTo converts from the versioned ResourceProviderResource resource to version-agnostic datamodel.
+func (src *ResourceProviderResource) ConvertTo() (v1.DataModelInterface, error) {
+	dst := &datamodel.ResourceProvider{
+		BaseResource: v1.BaseResource{
+			TrackedResource: v1.TrackedResource{
+				ID:       to.String(src.ID),
+				Name:     to.String(src.Name),
+				Type:     to.String(src.Type),
+				Location: to.String(src.Location),
+				Tags:     to.StringMap(src.Tags),
+			},
+		},
+	}
+
+	// Note: we omit SystemData and Tags for this type. They cannot be specified by the user.
+
+	dst.Properties = datamodel.ResourceProviderProperties{
+		Locations: map[string]datamodel.ResourceProviderLocation{},
+	}
+
+	for name, location := range src.Properties.Locations {
+		dst.Properties.Locations[name] = fromResourceProviderLocation(location)
+	}
+
+	for _, rt := range src.Properties.ResourceTypes {
+		dst.Properties.ResourceTypes = append(dst.Properties.ResourceTypes, fromResourceType(rt))
+	}
+
+	return dst, nil
+}
+
+// ConvertFrom converts from version-agnostic datamodel to the versioned ResourceProviderResource resource.
+func (dst *ResourceProviderResource) ConvertFrom(src v1.DataModelInterface) error {
+	dm, ok := src.(*datamodel.ResourceProvider)
+	if !ok {
+		return v1.ErrInvalidModelConversion
+	}
+
+	dst.ID = to.Ptr(dm.ID)
+	dst.Name = to.Ptr(dm.Name)
+	dst.Type = to.Ptr(ResourceProviderResourceType)
+	dst.Location = to.Ptr(dm.Location)
+
+	// Note: we omit SystemData and Tags for this type. They cannot be specified by the user.
+
+	dst.Properties = &ResourceProviderProperties{
+		ProvisioningState: to.Ptr(ProvisioningState(dm.InternalMetadata.AsyncProvisioningState)),
```

Comment:
I feel like this question is asked every time I write a converter 😆. 

- Provisioning state is read-only and cannot be set by the user. 
- I tend to include provisioning state in synchronous APIs for consistency since we show it in the CLI.
- I tend to default provisioning state -> succeeded because it's a simple design that reports the correct value. For cases where async processing needs to occur, the provisioning state will be non-nil.